src/main/java/com/zendesk/maxwell/producer/InflightMessageList.java

```diff
@@ -0,0 +1,64 @@
+package com.zendesk.maxwell.producer;
+/* respresents a list of inflight messages -- stuff being sent over the
+   network, that may complete in any order.  Allows for only bumping
+   the binlog position upon completion of the oldest outstanding item.
+
+   Assumes .addInflight(position) will be call monotonically.
+   */
+
+import com.zendesk.maxwell.BinlogPosition;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.ListIterator;
+
+public class InflightMessageList {
+	class InflightMessage {
+		public final BinlogPosition position;
+		public boolean isComplete;
+		InflightMessage(BinlogPosition position) {
+			this.position = position;
+			this.isComplete = false;
+		}
+	}
+
+	private LinkedList<InflightMessage> list;
+	private HashMap<String, InflightMessage> hash;
+
+	public InflightMessageList() {
+		this.list = new LinkedList<>();
+		this.hash = new HashMap<>();
```

Comment:
Yeah, ok.  The whole thing with modifying iterators and such is pretty confusing but it looks like that'd work.